matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/network/RetrofitExtensions.kt

```diff
@@ -91,10 +89,11 @@ private fun toFailure(errorBody: ResponseBody?, httpCode: Int, globalErrorReceiv
                 matrixError.code == MatrixError.M_CONSENT_NOT_GIVEN && !matrixError.consentUri.isNullOrBlank() -> {
                     globalErrorReceiver?.handleGlobalError(GlobalError.ConsentNotGivenError(matrixError.consentUri))
                 }
-                httpCode == HttpURLConnection.HTTP_UNAUTHORIZED && /* 401 */
-                        matrixError.code == MatrixError.M_UNKNOWN_TOKEN                                        -> {
-                    globalErrorReceiver?.handleGlobalError(GlobalError.InvalidToken(matrixError.isSoftLogout.orFalse()))
-                }
+// If using refresh tokens we won't want this as 401 will just mean refresh your token.
+//                httpCode == HttpURLConnection.HTTP_UNAUTHORIZED && /* 401 */
+//                        matrixError.code == MatrixError.M_UNKNOWN_TOKEN                                        -> {
+//                    globalErrorReceiver?.handleGlobalError(GlobalError.InvalidToken(matrixError.isSoftLogout.orFalse()))
+//                }
```

Comment:
How softLogout is managed now?

Reply:
If refreshTokenAuth is disable this would continue to function as it was before. If it is enable this error not just indicates we need to refresh. If this error gets returned from the refresh endpoint *then* respect it as before and look for the `soft_logout` as before. It's details in the [spec](https://github.com/matrix-org/matrix-doc/blob/main/proposals/2918-refreshtokens.md) linked in the issue.


> If a client uses an access token that has expired, the server will respond with an M_UNKNOWN_TOKEN error, preferably with the soft_logout parameter set to true to improve the user experience in legacy clients. Thus, if a client receives an M_UNKNOWN_TOKEN error, and it has a refresh token available, it should no longer assume that it has been logged out, and instead attempt to refresh the token. If the client was in fact logged out, then the server will respond with an M_UNKNOWN_TOKEN error to the token refresh request, possibly with the soft_logout parameter set.